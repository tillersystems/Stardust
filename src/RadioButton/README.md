# RadioButton

### Usage

A Radio represents a single value from multiple options, so a Radio should be used with at least another one.

```jsx
import { RadioButton } from '@tillersystems/stardust';
```

<!-- STORY -->

<!-- PROPS -->

### Example

```jsx
import { RadioButton } from '@tillersystems/stardust';

render() {
  return <RadioButton selectedValue={selectedValue} key="apple" id="apple" name="fruit" value="apple" onChange={this.handleChange}>apple</RadioButton>
}
```
