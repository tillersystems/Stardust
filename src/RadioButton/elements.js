import styled, { css } from 'styled-components';

/**
 * Wrapper for the actual check box and its label.
 *
 * @return {jsx}
 */
export const Wrapper = styled.div`
  display: flex;
  align-items: center;

  cursor: pointer;

  padding: 0.6rem;

  ${({ disabled }) =>
    disabled &&
    css`
      cursor: not-allowed;
    `};
`;

/**
 * Container of the actual check box.
 *
 * @return {jsx}
 */
export const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;

  width: 1.6rem;
  height: 1.6rem;

  border: 1px solid ${({ theme: { palette } }) => palette.lightGrey};
  border-radius: 50%;

  background: linear-gradient(
    0deg,
    ${({ theme: { palette } }) => palette.paleGrey} 0%,
    ${({ theme: { palette } }) => palette.white} 100%
  );
  box-shadow: inset 0 2px 0 0 ${({ theme: { palette } }) => palette.whiteOpacity(0.05)},
    0 1px 1px 0 ${({ theme: { palette } }) => palette.paleGrey};

  transition: background-color ease 0.25s;

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.4;
    `};

  ${({ checked }) =>
    checked &&
    css`
      border: 1px solid ${({ theme: { palette } }) => palette.primary.dark};
      background: ${({ theme: { palette } }) => palette.primary.default}
        linear-gradient(
          0deg,
          ${({ theme: { palette } }) => palette.whiteOpacity(0)} 0%,
          ${({ theme: { palette } }) => palette.whiteOpacity(0.1)} 100%
        );
      box-shadow: inset 0 0.2rem 0 0 ${({ theme: { palette } }) => palette.whiteOpacity(0.1)};
      transition: border ease 0.25s;
      position: relative;
    `};

  &:after {
    ${({ checked }) =>
      checked &&
      css`
        content: '';
        position: absolute;
        left: 50%;
        top: 50%;
        width: 0.6rem;
        height: 0.6rem;
        transform: translate3d(-50%, -50%, 0);
        background-color: ${({ theme: { palette } }) => palette.white};
        border-radius: 50%;
      `};
  }
`;

/**
 * Label of the check box.
 *
 * @return {jsx}
 */
export const Label = styled.label`
  display: block;

  cursor: pointer;

  position: relative;

  margin-left: ${({ theme: { dimensions } }) => dimensions.small};
  color: ${({ theme: { palette } }) => palette.spaceGrey};

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.4;
      cursor: not-allowed;
    `};

  ${({ checked }) =>
    checked &&
    css`
      color: ${({ theme: { palette } }) => palette.darkBlue};
    `};
`;
