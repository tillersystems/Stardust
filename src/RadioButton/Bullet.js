import React from 'react';
import PropTypes from 'prop-types';

import { StyledRadio } from './elements';

const RadioBullet = ({ isChecked, disabled }) => (
  <StyledRadio
    className={isChecked ? 'checked' : ''}
    disabled={disabled}
    data-testid="styled-radio"
  />
);

RadioBullet.propTypes = {
  isChecked: PropTypes.bool,
  disabled: PropTypes.bool,
};

RadioBullet.defaultProps = {
  isChecked: false,
  disabled: false,
};

export default RadioBullet;
