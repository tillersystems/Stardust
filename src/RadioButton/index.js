import React from 'react';
import PropTypes from 'prop-types';

import { Wrapper, Container, Label } from './elements';

/**
 * Defines a radio button component.
 *
 * @return {jsx}
 */
const RadioButton = ({
  children,
  className,
  disabled,
  id,
  name,
  onChange,
  selectedValue,
  value,
}) => {
  const checked = selectedValue === value;
  return (
    <Wrapper disabled={disabled} onClick={() => onChange(value)} className={className}>
      <Container checked={checked} disabled={disabled}>
        <input
          defaultChecked={checked}
          type="radio"
          name={name}
          value={value}
          id={id}
          disabled={disabled}
        />
      </Container>
      <Label htmlFor={id} checked={checked} disabled={disabled}>
        {children}
      </Label>
    </Wrapper>
  );
};

const { bool, func, node, string } = PropTypes;

/** Prop types. */
RadioButton.propTypes = {
  children: node,
  className: string,
  disabled: bool,
  id: string,
  name: string,
  onChange: func,
  selectedValue: string,
  value: string,
};

/** Default props. */
RadioButton.defaultProps = {
  children: null,
  className: '',
  disabled: false,
  id: null,
  name: null,
  onChange: () => {},
  selectedValue: null,
  value: '',
};

export default RadioButton;
