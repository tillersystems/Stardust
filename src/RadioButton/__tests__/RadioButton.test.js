import React from 'react';
import 'jest-styled-components';

import RadioButton from '..';

describe('<RadioButton />', () => {
  it('should render without a problem', () => {
    const { container } = render(<RadioButton id="test">test</RadioButton>);

    expect(container).toMatchSnapshot();
  });

  it('should render without a problem when radio is selected and enabled', () => {
    const { container } = render(
      <RadioButton id="test" value="apple" selectedValue="apple">
        test
      </RadioButton>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without a problem when radio is not selected and enabled', () => {
    const { container } = render(
      <RadioButton id="test" value="apple" selectedValue="apricot">
        test
      </RadioButton>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without a problem when radio is selected  and disabled', () => {
    const { container } = render(
      <RadioButton id="test" value="apple" selectedValue="apple" disabled>
        test
      </RadioButton>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without a problem when radio is not selected and disabled', () => {
    const { container } = render(
      <RadioButton id="test" value="apple" selectedValue="apricot" disabled>
        test
      </RadioButton>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
