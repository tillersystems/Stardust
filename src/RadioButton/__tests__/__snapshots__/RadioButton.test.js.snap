// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton /> should render without a problem 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg, hsl(240,33%,99%) 0%, hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05), 0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
}

.c2 {
  display: block;
  cursor: pointer;
  position: relative;
  margin-left: 0.7rem;
  color: hsl(207,13%,45%);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        id="test"
        type="radio"
        value=""
      />
    </div>
    <label
      class="c2"
      for="test"
    >
      test
    </label>
  </div>
</div>
`;

exports[`<RadioButton /> should render without a problem when radio is not selected and disabled 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
  cursor: not-allowed;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg, hsl(240,33%,99%) 0%, hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05), 0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
  opacity: 0.4;
}

.c2 {
  display: block;
  cursor: pointer;
  position: relative;
  margin-left: 0.7rem;
  color: hsl(207,13%,45%);
  opacity: 0.4;
  cursor: not-allowed;
}

<div
  class="c0"
  disabled=""
>
  <div
    class="c1"
    disabled=""
  >
    <input
      disabled=""
      id="test"
      type="radio"
      value="apple"
    />
  </div>
  <label
    class="c2"
    disabled=""
    for="test"
  >
    test
  </label>
</div>
`;

exports[`<RadioButton /> should render without a problem when radio is not selected and enabled 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg, hsl(240,33%,99%) 0%, hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05), 0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
}

.c2 {
  display: block;
  cursor: pointer;
  position: relative;
  margin-left: 0.7rem;
  color: hsl(207,13%,45%);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      id="test"
      type="radio"
      value="apple"
    />
  </div>
  <label
    class="c2"
    for="test"
  >
    test
  </label>
</div>
`;

exports[`<RadioButton /> should render without a problem when radio is selected  and disabled 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
  cursor: not-allowed;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg, hsl(240,33%,99%) 0%, hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05), 0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
  opacity: 0.4;
  border: 1px solid #1873a0;
  background: hsl(200,74%,46%) linear-gradient( 0deg, hsla(0,100%,100%,0) 0%, hsla(0,100%,100%,0.1) 100% );
  box-shadow: inset 0 0.2rem 0 0 hsla(0,100%,100%,0.1);
  -webkit-transition: border ease 0.25s;
  transition: border ease 0.25s;
  position: relative;
}

.c1:after {
  content: '';
  position: absolute;
  left: 50%;
  top: 50%;
  width: 0.6rem;
  height: 0.6rem;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
  background-color: hsl(0,100%,100%);
  border-radius: 50%;
}

.c2 {
  display: block;
  cursor: pointer;
  position: relative;
  margin-left: 0.7rem;
  color: hsl(207,13%,45%);
  opacity: 0.4;
  cursor: not-allowed;
  color: hsl(213,17%,20%);
}

<div
  class="c0"
  disabled=""
>
  <div
    class="c1"
    disabled=""
  >
    <input
      checked=""
      disabled=""
      id="test"
      type="radio"
      value="apple"
    />
  </div>
  <label
    class="c2"
    disabled=""
    for="test"
  >
    test
  </label>
</div>
`;

exports[`<RadioButton /> should render without a problem when radio is selected and enabled 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg, hsl(240,33%,99%) 0%, hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05), 0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
  border: 1px solid #1873a0;
  background: hsl(200,74%,46%) linear-gradient( 0deg, hsla(0,100%,100%,0) 0%, hsla(0,100%,100%,0.1) 100% );
  box-shadow: inset 0 0.2rem 0 0 hsla(0,100%,100%,0.1);
  -webkit-transition: border ease 0.25s;
  transition: border ease 0.25s;
  position: relative;
}

.c1:after {
  content: '';
  position: absolute;
  left: 50%;
  top: 50%;
  width: 0.6rem;
  height: 0.6rem;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
  background-color: hsl(0,100%,100%);
  border-radius: 50%;
}

.c2 {
  display: block;
  cursor: pointer;
  position: relative;
  margin-left: 0.7rem;
  color: hsl(207,13%,45%);
  color: hsl(213,17%,20%);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      checked=""
      id="test"
      type="radio"
      value="apple"
    />
  </div>
  <label
    class="c2"
    for="test"
  >
    test
  </label>
</div>
`;
